src/common.rs

```diff
@@ -0,0 +1,173 @@
+        write!(f, "{:?}", self.as_ref())
+    }
+}
+
+impl Display for BoxedErr {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        write!(f, "{}", self.as_ref())
+    }
+}
+
+impl Error for BoxedErr {
+    fn description(&self) -> &str {
+        self.as_ref().description()
+    }
+}
+
+/// The component type for futures of `T`.
+pub type FutureComponent<T> = Box<Future<Item = T, Error = BoxedErr> + Send + Sync>;
+
+impl<T> Component for FutureComponent<T>
+    where T: Send + Sync + 'static
+{
+    type Storage = DenseVecStorage<FutureComponent<T>>;
+}
+
+/// A resource you can use to store errors that occurred outside of
+/// the ECS but were catched inside, therefore should be handled by the user.
+#[derive(Debug)]
+pub struct Errors {
+    /// The collection of errors.
+    pub errors: Vec<BoxedErr>,
+}
+
+impl Errors {
+    /// Creates a new instance of `Errors`.
+    pub fn new() -> Self {
+        Errors { errors: Vec::new() }
+    }
+
+    /// Add an error to the error `Vec`.
+    pub fn add(&mut self, error: BoxedErr) {
+        self.errors.push(error);
+    }
+
+    /// Prints all errors and exits. Useful for debugging.
+    pub fn print_and_exit(&mut self) {
+        use std::io::stderr;
+        use std::process::exit;
+
+        if self.errors.is_empty() {
+            return;
+        }
+
+        writeln!(&mut stderr(),
+                 "Exiting program because of {} errors...",
+                 self.errors.len()).unwrap();
+
+        for error in self.errors.drain(..) {
+            let error = error.as_ref();
+
+            writeln!(&mut stderr(),
+                     "-----\n\tDescription: {}\n\tDebug: {:?}\n\tDisplay: {}",
+                     error.description(),
+                     error,
+                     error).unwrap();
+        }
+
+        exit(1);
+    }
+}
+
+/// A system which merges `Ready` futures into the persistent storage.
+/// Please note that your `World` has to contain a component storage
+/// for `T`.
+///
+/// In case of an error, it will be added to the `Errors` resource.
+pub struct Merge<T> {
+    asset_type: PhantomData<T>,
+}
+
+impl<T> Merge<T> {
+    /// Creates a new merge system.
+    pub fn new() -> Self {
+        Merge {
+            asset_type: PhantomData,
+        }
+    }
+}
+
+impl<'a, T> System<'a> for Merge<T>
+    where T: Component + Send + Sync + 'static
+{
+    type SystemData = (Entities<'a>,
+                       FetchMut<'a, Errors>,
+                       WriteStorage<'a, FutureComponent<T>>,
+                       WriteStorage<'a, T>);
+
+    fn run(&mut self, data: Self::SystemData) {
+        let (entities, mut errors, mut future, mut pers) = data;
+
+        let mut delete = Vec::new();
+
+        for (e, future) in (&*entities, &mut future).join() {
+            match future.poll() {
+                Ok(Async::Ready(x)) => {
+                    pers.insert(e, x);
+                    delete.push(e);
```

Comment:
Because `future` (the storage) is borrowed already. There's no way of removing from the storage you're iterating over.